### ISSUE TITLE
Failure to create a temporary password

### ISSUE THREAD

**Issue by openmindead** [electrickite/luks-tpm2#1]
Good day.
First of all, thank you for your work, you've done a great job that looks very promising!

In general the tool works fine (init, reset), but not a `temp` or `replace` commands:
```
sudo luks-tpm2 temp

Usealing keyfile...
ERROR: 
Load Object Failed ! ErrorCode: 0x902

ERROR: Unable to run tpm2_load
ERROR: Error opening file "/root/keyfs/object.context" due to error: No such file or directory
ERROR: Unable to run tpm2_unseal
Preparing to set a temporary LUKS passphrase for /dev/nvme0n1p5...
Failed to open key file.
A temporary passphrase was not set
```

This is what I get when pacman hook is executed:
```
[2018-11-29 13:39] [ALPM] running 'luks-tpm2.hook'...
[2018-11-29 13:39] [ALPM-SCRIPTLET] Usealing keyfile...
[2018-11-29 13:39] [ALPM-SCRIPTLET] ERROR: 
[2018-11-29 13:39] [ALPM-SCRIPTLET] Load Object Failed ! ErrorCode: 0x902
[2018-11-29 13:39] [ALPM-SCRIPTLET] 
[2018-11-29 13:39] [ALPM-SCRIPTLET] ERROR: Unable to run tpm2_load
[2018-11-29 13:39] [ALPM-SCRIPTLET] ERROR: Error opening file "/root/keyfs/object.context" due to error: No such file or directory
[2018-11-29 13:39] [ALPM-SCRIPTLET] ERROR: Unable to run tpm2_unseal
[2018-11-29 13:39] [ALPM-SCRIPTLET] Preparing to set a temporary LUKS passphrase for /dev/nvme0n1p5...
[2018-11-29 13:39] [ALPM-SCRIPTLET] /usr/bin/luks-tpm2: line 94: /dev/tty: No such device or address
[2018-11-29 13:39] [ALPM-SCRIPTLET] A temporary passphrase was not set
```

My settings in /etc/default/luks-tpm2, the rest is unchanged:
```
SEALED_KEY_PUBLIC="/boot/efi/keyfile.pub"
SEALED_KEY_PRIVATE="/boot/efi/keyfile.priv"
PARENT_HANDLE="0x81000003"
ROOT_DEVICE="/dev/nvme0n1p5"
```

What should be in /root/keyfs?

My OS is Manjaro Linux.

**Comment by electrickite** [comment]
Error 0x902 is:

    name: TPM2_RC_OBJECT_MEMORY
    description: out of memory for object contexts

In my experience, what this usually means is that the TPM device is being accessed directly, rather than using a resource manager. By default, `luks-tpm2` should use the in-kernel resource manager `/dev/tpmrm0`.

It would be helpful if you could provide:

  1. `luks-tpm2` version
  2. full contents of `/etc/default/luks-tpm2`
  3. kernel version `uname -a`
  4. tpm2 tools version `tpm2_getcap -v`

**Comment by electrickite** [comment]
Something else to try: does the `temp` action work if you reboot your machine and do not run any other `luks-tpm2` commands?

**Comment by openmindead** [comment]
Sure:

1. I installed luks-tpm2 from AUR, its version is 0.6.0-1.

2. `cat /etc/default/luks-tpm2`:

```
# Configuration file for luks-tpm2
#
# Commented variables show default values.

# Common settings
#
# TMPFS_MOUNT          Mount point for the tmpfs file system
# SEALED_KEY_PUBLIC    Absolute path to public portion of the sealed key
# SEALED_KEY_PRIVATE   Absolute path to private portion of the sealed key
# PARENT_HANDLE        Handle of the parent key when sealing data on disk
# NVRAM_INDEX          Index of NVRAM area to store key
# KEY_SIZE             Size in bytes of the generated key
# TPM_KEY_SLOT         LUKS slot number for the TPM-managed key
# RESET_KEY_SLOT       LUKS slot number for temporary reset passphrase
# PCRS                 TPM2-Tools PCR bank selection list used to seal key
# UNSEAL_PCRS          TPM2-Tools PCR bank selection list used to unseal key
# TPM2TOOLS_ENV_TCTI   TPM2-Tools TCTI identifier
# ROOT_DEVICE          The LUKS block device path
# ACTION               The default command action

#TMPFS_MOUNT="/root/keyfs"
#SEALED_KEY_PUBLIC="/boot/keyfile.pub"
#SEALED_KEY_PRIVATE="/boot/keyfile.priv"
#PARENT_HANDLE="0x81000001"
#NVRAM_INDEX=""
#KEY_SIZE=32
#TPM_KEY_SLOT=1
#RESET_KEY_SLOT=2
#PCRS="sha1:0,2,4,7"
#UNSEAL_PCRS="sha1:0,2,4,7"
#TPM2TOOLS_ENV_TCTI="device:/dev/tpmrm0"
SEALED_KEY_PUBLIC="/boot/efi/keyfile.pub"
SEALED_KEY_PRIVATE="/boot/efi/keyfile.priv"
PARENT_HANDLE="0x81000003"
ROOT_DEVICE="/dev/nvme0n1p5"

```
3. `uname -a`:

> Linux vladimir-xmnbp 4.19.5-2-MANJARO #1 SMP PREEMPT Wed Nov 28 18:19:10 UTC 2018 x86_64 GNU/Linux

4. `tpm2_getcap -v`:

> tool="tpm2_getcap" version="3.1.3" tctis="dynamic" tcti-default=device dlclose=enabled


As for `temp` action, it never worked for me. First I tried it right after `init` command (just to see if everything works) -- and it didn't, then after a reboot -- with the same result. In fact I've been using `luks-tpm2` for a quite some time (2-3 days) and already have some kernel updates installed since the installation of your tool, and it always ended up with this error for `temp` action.

**Comment by openmindead** [comment]
OK it seems I'm beginning to get puzzled more and more. Initially I used your README to set this up, and here is what I did and what I got:

`sudo tpm2_listpersistent -T device:/dev/tpmrm0`

> persistent-handle[0]:0x81000001 key-alg:rsa hash-alg:sha256 object-attr:fixedtpm|fixedparent|sensitivedataorigin|userwithauth|noda|restricted|decrypt
> persistent-handle[1]:0x81000002 key-alg:rsa hash-alg:sha256 object-attr:fixedtpm|fixedparent|sensitivedataorigin|userwithauth|noda|restricted|sign
> persistent-handle[2]:0x81010001 key-alg:rsa hash-alg:sha256 object-attr:fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decrypt

The above are handlers which existed before I created the others as follows:

`sudo tpm2_createprimary -H o -g sha1 -G rsa -T device:/dev/tpmrm0`

> ObjectAttribute: 0x00030072
> 
> CreatePrimary Succeed ! Handle: 0x80000000

`sudo tpm2_createprimary -H o -g sha256 -G rsa -T device:/dev/tpmrm0`

> ObjectAttribute: 0x00030072
> 
> CreatePrimary Succeed ! Handle: 0x80000001

`sudo tpm2_evictcontrol -A o -H 0x80000000 -S 0x81000003 -T device:/dev/tpmrm0`

> persistentHandle: 0x81000003

`sudo tpm2_evictcontrol -A o -H 0x80000001 -S 0x81000004 -T device:/dev/tpmrm0`

> persistentHandle: 0x81000004

So as you see I've created 2 handlers (just in case, for no reason) and used the first one for `init`.
Now I tried the second one - `0x81000004`, and the `temp` command worked fine:

`sudo luks-tpm2 -p /boot/keyfile -H 0x81000004 /dev/nvme0n1p5 init`

> Initializing LUKS TPM key for /dev/nvme0n1p5
> WARNING: This will permanently delete the key in slot 1!
> Do you wish to proceed? [Y/n]y
> Enter any existing LUKS passphrase: 
> Generating new LUKS key...
> Removing existing key from slot 1...
> Adding new key to slot 1...
> Sealing keyfile with the TPM...

`sudo luks-tpm2 temp`

> Usealing keyfile...
> Preparing to set a temporary LUKS passphrase for /dev/nvme0n1p5...
> Enter new passphrase for key slot: 
> Verify passphrase: 
> After booting into the current kernel, run
>   luks-tpm2 reset
> to generate a new LUKS key and remove this temporary key

However, after reboot I have the same error 0x902 when trying to use `temp` or `replace` commands.

**Comment by openmindead** [comment]
The following sequence works:
`init` -> reboot -> choosing wrong handler in bootloader options -> booting with entering permanent keyphrase from slot 0 -> `temp` command works. However, this is not meant to work this way, is it?

**Comment by electrickite** [comment]
@openmindead Are you using the tpm2 mkinitcpio hook? If so, what are your kernel command line parameters for that hook?

**Comment by electrickite** [comment]
It would also be helpful to know your TSS version: `pacman -Qi tpm2-tss`

**Comment by openmindead** [comment]
`pacman -Qi tpm2-tss`
> Name            : tpm2-tss
> Version         : 2.1.0-1
> Description     : TPM (Trusted Platform Module) 2.0 Software Stack (TSS)
> Architecture    : x86_64
> URL             : https://github.com/tpm2-software/tpm2-tss
> Licenses        : BSD
> Groups          : None
> Provides        : None
> Depends On      : libgcrypt
> Optional Deps   : None
> Required By     : tpm2-tools
> Optional For    : None
> Conflicts With  : None
> Replaces        : None
> Installed Size  : 1420.00 KiB
> Packager        : Unknown Packager
> Build Date      : Sat 24 Nov 2018 15:54:15 +10
> Install Date    : Sat 24 Nov 2018 16:05:12 +10
> Install Reason  : Explicitly installed
> Install Script  : No
> Validated By    : None

Kernel command line options related to tpm2 are:
> tpmkey=/dev/nvme0n1p6:/keyfile:0x81000004 tpmpcr=sha1:0,2,4,7

nvme0n1p6 is where keyfiles (priv and pub) are located.

**Comment by electrickite** [comment]
Turns out there was a regression in the TCTI configuration for tpm2-tools: https://github.com/tpm2-software/tpm2-tools/issues/1233

Because of this, TSS was falling back to the default `/dev/tpm0` device, which does not use a resource manager. This was likely causing the out-of-memory errors.

Try the latest luks-tpm2 0.7.0 and see if it resolves the issue. (Please reboot your computer after installing to ensure the TPM context is flushed)

**Comment by openmindead** [comment]
@electrickite I deleted luks-tpm2 and mkinitcpio-tpm2-encrypt and reinstalled them both. Unfortunately, nothing changed in behavior of the tool.
It successfully inits sealing the key. It resets if needed. But temp and replace only work if key is sealed at the moment of command execution. As I described before:
The following works:
`luks-tpm2 init` -> reboot -> intentional choosing wrong handler in bootloader options -> booting with entering permanent keyphrase from slot 0 -> temp/replace command works (maybe because the key is sealed at that moment?)
The following leads to failure:
`luks-tpm2 init` -> reboot -> choosing with correct handler in bootloader options -> booting with tpm2 hook unsealing the key -> temp/replace fails to unseal (maybe because the key is already unsealed by initcpio hook? -- I'm not sure).

**Comment by electrickite** [comment]
You have version 0.7.0? `luks-tpm2 -v`

**Comment by openmindead** [comment]
Yeah of course:
`luks-tpm2 -v`

> luks-tpm2 0.7.0

**Comment by electrickite** [comment]
Alright, I managed to reproduce the error on my machine and resolved it here. Please update mkinitcpio-tpm2-encrypt to version 1.2.2-1 and try one more time.

The same tpm2-tools TCTI configuration regression from upstream needed to be solved in the mkinitcpio hook. Without the fix in 1.2.2, the objects loaded during disk decryption were staying resident in TPM memory, causing the out of memory errors when additional objects were loaded by luks-tpm2.

**Comment by openmindead** [comment]
Yeah, it finally works! Tnank you. I'm glad I helped you to hunt down such not an obvious bug.
